[OpRefactor] Upgrade matrix methods

mariaschuld opened the pull request:
**Context:**

Implements the changes to the matrix representation.

**Description of the Change:**

* Turned `_matrix` into static method `compute_matrix(*params, **hyperparams)` which defines the canonical matrix. 
* Add function `expand_matrix` which differentiably re-expresses a matrix according to a global wire order. Also add tests.
* Change `.matrix` to `.matrix(wire_order=None)`, which calls `expand_matrix`. Add some non-trivial tests.
* Update methods, docstrings and tests for all operations.

**Benefits:**

We should be able to cache the matrix expansion (which needs some tricks, since arrays are non-hashable). The matrix representation now has a clear interpretation. All methods have examples.

**Downsides:**
There are many ways how to deal with operations where `compute_matrix` relies on values that we don't want to compute each time, I opted for one and discuss others.

**NOTE** Currently there is a test that does not pass on CI, but passes fine locally. I will investigate, but would still like to open for review!
![Screenshot from 2021-12-16 11-47-44](https://user-images.githubusercontent.com/23052011/146351378-f789b850-c476-46ce-9894-c0999a2897c3.png)

I will also check the docs properly once it looks like we are close to merging!

mariaschuld commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -53,9 +52,9 @@ def num_params(self):
-        return cls.matrix
+    @staticmethod
+    def _matrix(*params):
+        return np.array([[INV_SQRT2, INV_SQRT2], [INV_SQRT2, -INV_SQRT2]])
```
Here and below: matrix is not defined in a variable called by a classmethod, but is defined in the static method. If necessary, subclasses can use classmethods to have `cls` available, but it shouldn't be encouraged so that we get as little dependency on what is available in the class as possible...

albi3ro commented on pennylane/grouping/utils.py:
```diff
@@ -523,7 +523,7 @@ def pauli_word_to_matrix(pauli_word, wire_map=None):
         if wire_label in pauli_word.wires.labels:
             op_idx = pauli_word.wires.labels.index(wire_label)
-            pauli_mats[wire_idx] = getattr(qml, pauli_names[op_idx]).matrix
+            pauli_mats[wire_idx] = getattr(qml, pauli_names[op_idx]).compute_matrix()
```
Doesn't `compute_matrix` need parameters?

mariaschuld replied:
Not for Paulis, so here we are ok. But maybe you're right, it may not me safe. I am not entirely sure what to do here.

mariaschuld replied:
I added a comment, and unless someone changes this function it should be safe!

albi3ro commented on pennylane/operation.py:
```diff
@@ -300,48 +300,149 @@ def hash(self):
+
     @classmethod
-    def _matrix(cls, *params):
+    def compute_matrix(cls, *params, wires=None, wire_order=None, **hyperparams):
```
I'm just wondering if it makes sense to take `params` as a single iterable, instead of an unpacked arbitrary sized list.  Then the other variables can be passed as arguments instead of limiting them to only keyword arguments: `op.compute_matrix(op.parameters, op.wires, wire_order, op.hyperparameters)`.

josh146 replied:
@mariaschuld from reading this, it looks like it makes a lot of sense to simply 'merge' `_matrix` and `compute_matrix`, and make `compute_matrix` a static method?

albi3ro commented on pennylane/operation.py:
```diff
@@ -300,48 +300,149 @@ def hash(self):
+        # intersecting this case here because it is common
+        # and we want to avoid casting to Wires unless necessary
+        if wire_order is None:
+            # TODO: cast to interface?
```
Shouldn't `_matrix` handle casting to the proper interface? `_matrix` would need to use the interface to construct it in a differentiable way, so it should already be converted?

josh146 replied:
@albi3ro This is only the case for parametrized matrices, as the type can be inferred from the parameters. For non-parametrized matrices, casting will need to be done manually.

mariaschuld replied:
Actually, I think this is responsible for my last failing tests in backprop devices' gradients...do you know where the casting is currently done, @josh146?

josh146 commented on doc/releases/changelog-dev.md:
```diff
@@ -12,5 +12,17 @@
+* The `matrix` representation has been modified to be a method that accepts a 
+  `wire_order` argument and calculate the correct numerical representation 
+  with respect to that ordering. Internally, `op.matrix()` calls the new static 
+  version `compute_matrix()`, which in turn derives its information from the existing 
+  `_matrix()` method which defines the "base matrix" which uses the operator's wire order.
```
Can we remove `_matrix()`? It seems superfluous, and hard to document!

I must admit, I assumed that `compute_matrix` was the replacement for `_matrix` 🙂

mariaschuld replied:
The problem is this: we want 1) a method that is easy to call and dynamic, and respects the wire order (`matrix(wire_order=None)`), 2) a method that is static, cachable and respects the wire order (`compute_matrix(..., wire_order=None)`). But then we need a place where the base matrix can be defined. That would be `_matrix`, and it should not be called, it is just a data structure.

Do you think that is too complicated? 

My worry about merging `compute_matrix` and `_matrix` is that the former contains a lot of logic to expand the matrix. Everyone defining an op would have to reproduce the logic.

josh146 commented on pennylane/fourier/utils.py:
```diff
@@ -67,7 +67,8 @@ def get_spectrum(op, decimals):
             matrix = g
         elif hasattr(g, "matrix"):
-            matrix = g.matrix
+            # TODO [Maria]: if generators are class instances, change to matrix()
+            matrix = g.compute_matrix()
```
This whole if statement will be simplified with the generator change coming up 🙌

josh146 replied:
wait, shouldn't this take arguments?

mariaschuld replied:
Yes and no. I think at the moment all generators in the code base are Paulis and don't take args? 

This is a pattern I want to avoid in future: automatically calling class methods. Is there any hotfix we can use for now?

josh146 replied:
> all generators in the code base are Paulis and don't take args?

Oh of course!

mariaschuld replied:
I added a better comment

josh146 commented on pennylane/operation.py:
```diff
@@ -300,48 +300,149 @@ def hash(self):
 
         Returns:
-            array: matrix representation
+            tensor-like: matrix representation, or None if not defined
```
```suggestion
            tensor_like or None: matrix representation, or None if not defined
```

josh146 commented on pennylane/operation.py:
```diff
@@ -300,48 +300,149 @@ def hash(self):
+        )
+        axes = (list(range(n, 2 * n)), op_wire_pos)
+
+        # reshape op.matrix()
+        op_matrix_interface = qml.math.convert_like(base_matrix, I)
+        mat_op_reshaped = qml.math.reshape(op_matrix_interface, [2] * n * 2)
+        mat_tensordot = qml.math.tensordot(
+            mat_op_reshaped, qml.math.cast_like(I, mat_op_reshaped), axes
+        )
+
+        unused_idxs = [idx for idx in range(len(wire_order)) if idx not in op_wire_pos]
+        # permute matrix axes to match wire ordering
+        perm = op_wire_pos + unused_idxs
+        mat = qml.math.moveaxis(mat_tensordot, wire_order.indices(wire_order), perm)
+
+        mat = qml.math.reshape(mat, (2 ** len(wire_order), 2 ** len(wire_order)))
```
@mariaschuld I wonder if either of 

https://github.com/PennyLaneAI/pennylane/blob/1fd370d0ee2b14c3ddef9af6e5eccbf8784350a6/pennylane/utils.py#L307-L308

or 

https://github.com/PennyLaneAI/pennylane/blob/1fd370d0ee2b14c3ddef9af6e5eccbf8784350a6/pennylane/utils.py#L372-L373

can be used here?

Although, your implementation might be preferred since it uses `qml.math` :)

mariaschuld replied:
I copied it from Olivia's code to get a unitary. Will outsource it!

josh146 commented on pennylane/operation.py:
```diff
@@ -300,48 +300,149 @@ def hash(self):
+
+        n = len(wires)
+
+        # operator's wire positions relative to wire ordering
+        op_wire_pos = Wires(wire_order).indices(wires)
+
+        I = qml.math.reshape(
+            qml.math.eye(2 ** len(wire_order), like=interface), [2] * len(wire_order) * 2
+        )
+        axes = (list(range(n, 2 * n)), op_wire_pos)
```
I realize, from reading this: this is important internal logic that a user should _not_ overwrite or change!

As a result, it maybe doesn't make sense for a user to overwrite `compute_matrix()`.

Would it make sense to:

- Have a staticmethod `compute_matrix(parameters, hyperparameters)` that _simply returns the base matrix representation_.

- Have an instance method `matrix(self, wire_order)` that includes this wire ordering/expansion logic?

albi3ro replied:
I was thinking that code for rearranging the base matrix to have a different order could be extracted to a utility method, or even made a "transform".  Then this method would be much simpler.

josh146 replied:
Good idea!

mariaschuld replied:
Here is the problem @josh146: if the expansion logic is in `matrix()` only, we cannot cache it. Otherwise I'd be happy to merge the two! 

Maybe I can benchmark expansion and see how costly it is.

josh146 replied:
From our phone call just now: caching will be preferred on a staticmethod rather than a classmethod!

josh146 commented on pennylane/ops/qubit/matrix_ops.py:
```diff
@@ -212,25 +212,22 @@ def __init__(
         self._padding_right = 2 ** len(wires) - len(U) - self._padding_left
-        self._CU = None
 
+        # TODO[MARIA]: decide if it is too costly to compute this at creation
+        interface = qml.math.get_interface(self.U)
+        left_pad = qml.math.cast_like(qml.math.eye(self._padding_left, like=interface), 1j)
+        right_pad = qml.math.cast_like(qml.math.eye(self._padding_right, like=interface), 1j)
+        cu = qml.math.block_diag([left_pad, self.U, right_pad])
+
+        # make a QubitUnitary from the controlled matrix
+        new_params = list(params)
+        new_params[0] = cu
```
This seems like something that should be in `compute_matrix`?

mariaschuld replied:
whoops, this should be deleted!

josh146 reviewed (commented):
@mariaschuld have left some preliminary thoughts!

mariaschuld commented on pennylane/operation.py:
```diff
@@ -129,6 +129,99 @@
 from .utils import pauli_eigs
 
+
+def expand_matrix(base_matrix, wires, wire_order):
```
Josh and I thought of making this a transform, but I want it to be here in order to adapt it to future needs

mariaschuld replied:
Oh, interesting, I wasn't aware of this. So we have this logic in several places! A good reason to focus it somewhere...

The code seems to be slightly different, let me benchmark it!

josh146 replied:
> The code seems to be slightly different, let me benchmark it!

Yep, I wonder if the `utils` version might be more battle-hardened?

mariaschuld replied:
`expand` seems to be a bit faster, but the torch differentiation test fails (and the utils method does not claim to be differentiable). Still it would be neat to rely on existing functions...Maybe I can figure out what is going wrong

mariaschuld replied:
Ok, for now I added a note that the two should be merged some day.

josh146 replied:
We should also open an issue! this could make a very good tech assessment

josh146 replied:
let me know if you don't have time, and I'll will add it to my todo list tomorrow :)

mariaschuld commented on pennylane/ops/identity.py:
```diff
@@ -49,8 +49,9 @@ def label(self, decimals=None, base_label=None):
 
-    @classmethod
-    def _matrix(cls, *params):
+    # pylint:disable=unused-argument
```
I am unsure about this pattern, but I don't know how else to automate things

@albi3ro, shall I use `params, hyperparams` instead? The reason for the star is that I thought if a subclass overwrites this method with a different signature like `(eigvals, num_wires, basis_state)`, the highest-order method  `Operator.compute_matrix` will also work with this input because it does not specify a number of arguments.

josh146 replied:
I am in favour of `*params, **hyperparams`. We decided to do `qnode.construct(args, kwargs)` ages ago, and I really regret it --- it has led to a lot of bugs due to people forgetting to convert between the two. Plus, subclassing, as you say

josh146 commented on pennylane/operation.py:
```diff
@@ -129,6 +129,101 @@
+    )
+    axes = (list(range(n, 2 * n)), op_wire_pos)
+
+    # reshape op.matrix()
+    op_matrix_interface = qml.math.convert_like(base_matrix, I)
+    mat_op_reshaped = qml.math.reshape(op_matrix_interface, [2] * n * 2)
+    mat_tensordot = qml.math.tensordot(
+        mat_op_reshaped, qml.math.cast_like(I, mat_op_reshaped), axes
+    )
```
I'm curious why https://github.com/PennyLaneAI/pennylane/blob/4e49e9d62bd3358c28f3c156e9886ce0f789c807/pennylane/utils.py#L339-L352
has this logic?

josh146 replied:
Also I'm curious at the speed difference between the two :) It looks like this one might be faster

mariaschuld replied:
(see above)

josh146 commented on pennylane/operation.py:
```diff
@@ -300,48 +395,68 @@ def hash(self):
-        >>> qml.RY._matrix(0.5)
-        >>> array([[ 0.96891242+0.j, -0.24740396+0.j],
-                   [ 0.24740396+0.j,  0.96891242+0.j]])
+        Args:
```
as per our current conventions this should be moved above `Example`

mariaschuld replied:
Oh sorry, I forgot about that!

josh146 commented on pennylane/operation.py:
```diff
@@ -300,48 +395,68 @@ def hash(self):
-        in the computational basis.
+    # pylint:disable=unused-argument
+    @staticmethod
+    def compute_matrix(*params, **hyperparams):
```
I suppose it makes sense that this doesn't take `wires` right

mariaschuld replied:
Yes, and where a base matrix needs info like the number of wires, we just overwrite the signature!

mariaschuld replied:
One or two operations _do_ need wires. We have two options: a) pass it here by default (although it will be unused in most cases), or b) handle these cases separately. I opted for b) because I think we have to provide for deviations like this anyways.

josh146 commented on pennylane/operation.py:
```diff
@@ -300,48 +395,68 @@ def hash(self):
+            tensor-like or None: matrix representation
         """
-        return self._matrix(*self.parameters)
+        base_matrix = self.compute_matrix(*self.parameters, **self.hyperparameters)
+        return expand_matrix(base_matrix, wires=self.wires, wire_order=wire_order)
```
this looks really clean now!!

josh146 commented on pennylane/operation.py:
```diff
@@ -129,6 +129,101 @@
+            contain additional labels
+    """
+
+    shp = qml.math.shape(base_matrix)
+    if shp[0] != shp[1]:
+        raise ValueError(f"Expected square matrix to expand; got shape {shp}.")
```
This will break JIT I fear :(

mariaschuld replied:
Let me remove.

josh146 commented on pennylane/operation.py:
```diff
@@ -129,6 +129,101 @@
+    wire_order = Wires(wire_order)
+    wires = Wires(wires)
+
+    if Wires(wires) == Wires(wire_order):
+        return base_matrix
```
I wonder if this should be moved into `self.matrix`:

```python
def matrix(self):
    base_matrix = self.compute_matrix(*self.parameters, **self.hyperparameters)

    if self.wires == Wires(wire_order):
        return base_matrix

    return expand_matrix(base_matrix, wires=self.wires, wire_order=wire_order)
```

the advantage is that it avoids a call to `expand_matrix` if not required

mariaschuld replied:
Yes, I was wondering though that it has to be added every time we overwrite this method?

josh146 replied:
Good point 🤔 My main motivation here is to avoid calling `expand_matrix` if not needed, to avoid (a probably minor) overhead, and to avoid additional validation that occurs that might break the JIT.

mariaschuld replied:
I implemented the change, probably worth it!

josh146 commented on pennylane/operation.py:
```diff
@@ -1370,20 +1495,18 @@ def diagonalizing_gates(self):
-        subsystem it is defined for.
+            The wire_order argument is added for compatibility, but currently not implemented.
+
+        **Example:**
```
```suggestion
        **Example**
```

josh146 commented on pennylane/ops/qubit/observables.py:
```diff
@@ -235,6 +237,15 @@ def label(self, decimals=None, base_label=None):
+        basis_state = params[0]
+        m = np.zeros((2 ** len(basis_state), 2 ** len(basis_state)))
+        idx = int("".join(str(i) for i in basis_state), 2)
+        m[idx, idx] = 1
+        return m
```
I suppose the casting happens later now?

mariaschuld replied:
Oh good point! I added this because I saw it's lacking and easy to implement (famous last words). 

However I doubt here we can infer the framework, because `basis_state` is usually a list?

josh146 replied:
Oh so this is really interesting - `Projector` accepts `basis_state` as `tensor_like`, but in `__init__` it casts it first to a numpy array, and then to a list.

It really shouldn't be doing this, it should be leaving it as a tensor

josh146 replied:
Technically as well, `basis_state` is not a parameter but a hyperparameter, since it is not trainable?

josh146 replied:
I think I am to blame, I added this operator iirc!

josh146 reviewed (commented):
@mariaschuld this feels a lot cleaner now!! I think my only confusion stems from the difference between `qml.operation.expand_matrix` and `qml.utils.expand`. They seem to do the exact same thing, but the code looks very different. So I can't tell if there are potentially differences between the two.

mariaschuld commented on pennylane/operation.py:
```diff
@@ -129,6 +129,101 @@
+
+    If the wire order is identical to ``wires``, the original matrix gets returned:
+
+    >>> base_matrix = np.array([[1, 2, 3, 4],
+    >>>                         [5, 6, 7, 8],
+    >>>                         [9, 10, 11, 12],
+    >>>                         [13, 14, 15, 16]])
+    >>> expand_matrix(base_matrix, wires=[0, 2], wire_order=[0, 2])
```
Oh sorry!

mariaschuld commented on pennylane/_qubit_device.py:
```diff
@@ -761,13 +761,13 @@ def expval(self, observable, shot_range=None, bin_size=None):
-                ) from e
-
+                )
+            eigvals = self._asarray(eigvals, dtype=self.R_DTYPE)
```
We now assume that "not defined" is `None` instead of `NotImplementedError`.

Note that in standard situations like this we don't need to query if a representation is defined through some "registration", since if it is defined, we would like to use the result.

mariaschuld commented on pennylane/_qubit_device.py:
```diff
@@ -787,11 +787,11 @@ def var(self, observable, shot_range=None, bin_size=None):
+            if eigvals is None:
+                # no eigenvalue representation defined
+                raise ValueError(f"Cannot compute analytic variance of {observable.name}.")
+            eigvals = self._asarray(eigvals, dtype=self.R_DTYPE)
```
Same as above

mariaschuld commented on pennylane/_qubit_device.py:
```diff
@@ -829,11 +829,12 @@ def sample(self, observable, shot_range=None, bin_size=None):
+            if evals is None:
+                # the eigenvalue representation is not defined
+                raise ValueError(f"Cannot compute samples of {observable.name}.")
+            samples = evals[indices]
```
same as above

mariaschuld commented on pennylane/fourier/utils.py:
```diff
@@ -67,7 +67,9 @@ def get_spectrum(op, decimals):
-            matrix = g.matrix
+            # hotfix: this only works because all
+            # non-numerical generators are Paulis.
+            matrix = g.compute_matrix()
```
I want to add this comment so in the long run we find a better solution, and so developers looking for why their tests fail find it!

josh146 replied:
@mariaschuld this will be partially addressed in #2030!

mariaschuld commented on pennylane/operation.py:
```diff
@@ -300,48 +394,100 @@ def hash(self):
+            to ``op.hyperparameters``, and ``op`` is an instance of this operator.
+
+            If a subclass overwrites ``compute_matrix`` to use a custom signature,
+            this method has to be likewise overwritten.
```
Happy for suggestions to make this note clearer

mariaschuld replied:
Yes they could do that as well. It's a good point, because we could have `matrix` defined, but not `compute_matrix`. 

I think we should say a representation is defined if the `matrix` method is defined, and never rely on `compute_matrix`, which is more a convenience function or helper? This could be fixed by the "registration system"!

josh146 replied:
Yes! I'm in favour if it makes things easier for developers.

albi3ro replied:
What might be a reasonable situation where this could happen?

I see this as an avoid unless absolutely necessary situation.  Especially given how much Operator addition and maintenance is done by first-time contributors, we might not want to be suggesting something that isn't recommended except when absolutely necessary,

mariaschuld commented on pennylane/operation.py:
```diff
@@ -376,9 +522,12 @@ def _eigvals(cls, *params):
-        return np.linalg.eigvals(cls._matrix(*params))
+        mat = cls.compute_matrix(*params)
+        if mat is None:
+            return None
```
Need to propagate the possible case of an undefined matrix here!

josh146 replied:
Ohhh actually I guess that was quite nice in the old system --- you didn't need to worry about propagating `None`s, because a single `NotImplementedError` would break the chain for you automatically

mariaschuld replied:
Yes that's true! But then, we never knew where that error came from, so it was more opaque as well.

mariaschuld commented on pennylane/operation.py:
```diff
@@ -815,14 +964,16 @@ def inv(self):
 
         if self.inverse:
-            return qml.math.conj(qml.math.T(op_matrix))
+            canonical_matrix = qml.math.conj(qml.math.T(canonical_matrix))
```
I hope this will disappear with the new inverse system. It is annoying to replicate it when overwriting the method, and I am very confused which operator has to replicate it (just following the status quo below).

josh146 replied:
🤔

josh146 replied:
I can't stop thinking about this.

In the matrix case, we have `compute_matrix` (static, designed to be overridden) and `matrix` (instance method, contains post-processing logic).

In the generator case, we have `generator` (instance method, designed to be overridden), and `utils.get_generator(op)` (external function to add post-processing logic).

Different solutions for the same problem: how do we provide a stable API for developers to add a representation, while also abstracting away how that representation changes with the operator.

mariaschuld replied:
Still need to work this one out conclusively!!

mariaschuld commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -989,26 +1163,62 @@ def __init__(
-                np.eye(self._padding_left), PauliX.matrix, np.eye(self._padding_right)
-            )
+    @staticmethod
+    def compute_matrix(control_int, n_wires):
```
This is a case where we could do it many ways...we could pass `params, hyperparams` and do the entire computation in this method (only). Or we compute things like `control_int` in the init method and add it to the `hyperparams` (ignoring others). Here I overwrite the signature to take very specific inputs, and overwrite `matrix` to provide them.

mariaschuld commented on pennylane/ops/qubit/observables.py:
```diff
@@ -152,12 +166,33 @@ def num_params(self):
+        if not isinstance(H, coo_matrix):
             raise TypeError("Observable must be a scipy sparse coo_matrix.")
-        return A
+        return H
```
For now I keep this consistent with the status quo. In a follow-up PR, this method will convert the sparse to a dense matrix, and `sparse_matrix` will return the original one.

mariaschuld commented on pennylane/ops/qubit/observables.py:
```diff
@@ -235,6 +270,29 @@ def label(self, decimals=None, base_label=None):
+        """
+        m = np.zeros((2 ** len(basis_state), 2 ** len(basis_state)))
+        idx = int("".join(str(i) for i in basis_state), 2)
+        m[idx, idx] = 1
```
added this one from scratch, Projectors didn't have a matrix representation before

mariaschuld commented on pennylane/ops/qubit/parametric_ops.py:
```diff
@@ -59,10 +59,22 @@ class RX(Operation):
+
+        >>> qml.RX.compute_matrix(torch.tensor(0.5))
+        tensor([[0.9689+0.0000j, 0.0000-0.2474j],
+                [0.0000-0.2474j, 0.9689+0.0000j]])
```
I decided to use tensors in some of the examples. Ideally I could use different frameworks, but I was too lazy.

josh146 replied:
I do the same as well lol

mariaschuld commented on pennylane/ops/qubit/parametric_ops.py:
```diff
@@ -471,23 +547,49 @@ class MultiRZ(Operation):
-    def _matrix(cls, theta, n):
-        """Matrix representation of a MultiRZ gate.
+    @staticmethod
+    def compute_matrix(theta, n_wires):
```
This is another case where we overwrite, because info on the wires needs to be provided

josh146 replied:
@mariaschuld should this be updated to use `hyperaparams`?

mariaschuld replied:
Yes, this is one of the examples. Let me do it and see how far I get!

mariaschuld commented on pennylane/ops/qubit/qchem_ops.py:
```diff
@@ -89,15 +89,29 @@ def circuit(phi):
 
         mat = qml.math.diag([1, c, c, 1])
-        off_diag = qml.math.convert_like(np.diag([0, 1, -1, 0])[::-1].copy(), theta)
+        off_diag = qml.math.convert_like(np.diag([0, 1, -1, 0])[::-1].copy(), phi)
```
Renamed to be consistent with class docstring

mariaschuld commented on pennylane/templates/subroutines/qft.py:
```diff
@@ -70,17 +70,19 @@ def circuit_qft(basis_state):
     @functools.lru_cache()
-    def _matrix(cls, num_wires):
+    def compute_matrix(num_wires):
         dimension = 2 ** num_wires
```
Another reason to allow for custom signatures: some params and hyperparams are unhashable, in which case we may be happy about the flexibility

mariaschuld commented on pennylane/transforms/get_unitary_matrix.py:
```diff
@@ -139,25 +139,8 @@ def wrapper(*args, **kwargs):
-            U = qml.math.moveaxis(U_tensordot, wire_order.indices(wire_order), perm)
-
-            U = qml.math.reshape(U, ((2 ** n_wires, 2 ** n_wires)))
+            U = op.matrix(wire_order=wire_order)
```
I copied the code from `expand_matrix` from here, so can replace it now

mariaschuld commented on pennylane/utils.py:
```diff
@@ -85,7 +85,7 @@ def decompose_hamiltonian(H, hide_identity=False):
 
     for term in itertools.product(paulis, repeat=n):
-        matrices = [i._matrix() for i in term]
+        matrices = [i.compute_matrix() for i in term]
```
This assumes that all terms in the hamiltonian are ops that do not take operators!

mariaschuld commented on tests/ops/qubit/test_arithmetic_ops.py:
```diff
@@ -50,8 +50,8 @@ def test_label(op, label):
-class TestArithmetic:
-    """Tests the arithmetic operations."""
+class TestQubitCarry:
+    """Tests the QubitCarry operator."""
```
I couldn't help rearranging this class, so that we have one for each operation!

mariaschuld commented on tests/ops/qubit/test_arithmetic_ops.py:
```diff
@@ -200,4 +232,24 @@ def f(state):
+            ]
+        )
+        assert np.allclose(res_static, expected, atol=tol)
+        assert np.allclose(res_dynamic, expected, atol=tol)
```
I'm sorry I test both methods in one here, but I just couldn't get myself to duplicate every test. If needed, we can do that later?

mariaschuld commented on tests/ops/qubit/test_matrix_ops.py:
```diff
@@ -396,6 +416,22 @@ def circuit_pauli_x():
+            ]
+        )
+        # assert np.allclose(res_static, expected, atol=tol)
+        assert np.allclose(res_dynamic, expected, atol=tol)
```
TODO!

mariaschuld commented on tests/ops/qubit/test_observables.py:
```diff
@@ -106,9 +102,14 @@ def test_eigvals(self, obs, mat, eigs, tol):
+
+@pytest.mark.usefixtures("tear_down_hermitian")
+class TestHermitian:
+    """Test the Hermitian observable"""
```
Made a new class here to tidy up

mariaschuld commented on tests/ops/qubit/test_parametric_ops.py:
```diff
@@ -268,64 +268,74 @@ def test_controlled_phase_shift_decomp(self, phi, cphase_op):
-        assert np.allclose(qml.PhaseShift._matrix(0), np.identity(2), atol=tol, rtol=0)
-        assert np.allclose(qml.U1._matrix(0), np.identity(2), atol=tol, rtol=0)
+        assert np.allclose(qml.PhaseShift.compute_matrix(0), np.identity(2), atol=tol, rtol=0)
+        assert np.allclose(qml.U1.compute_matrix(0), np.identity(2), atol=tol, rtol=0)
```
Just adding tests for the base matrix here!

mariaschuld commented on tests/ops/test_identity.py:
```diff
@@ -0,0 +1,32 @@
+    res_dynamic = qml.Identity(wires=0).matrix()
+    expected = np.array([[1.0, 0.0], [0.0, 1.0]])
+    assert np.allclose(res_static, expected, atol=tol)
+    assert np.allclose(res_dynamic, expected, atol=tol)
```
I was confused to find the identity test in some random other file, so made a new one.

mariaschuld commented on tests/test_qubit_device.py:
```diff
@@ -498,10 +498,22 @@ def test_non_analytic_expval(self, mock_qubit_device_with_original_statistics, m
+
         with pytest.raises(ValueError, match="Cannot compute analytic expectations"):
-            dev.expval(qml.Hamiltonian([1.0], [qml.PauliX(0)]))
+            dev.expval(obs)
```
Changed this because it is safer to make an op that has no evals defined (since we may change this for Hamiltonian)

josh146 commented on pennylane/operation.py:
```diff
@@ -1350,20 +1501,25 @@ def diagonalizing_gates(self):
-    @property
-    def matrix(self):
-        r"""Matrix representation of the tensor operator
+    def matrix(self, wire_order=None):
+        r"""Matrix representation of the Tensor operator
```
I can't quite tell from the github view, but what class is this in/how come it is here as well as above?

mariaschuld replied:
This is the `Tensor` class, where a lot of different stuff happens :) Hopefully deleted soon!

Above method is in another class!

josh146 replied:
got it!

josh146 commented on pennylane/ops/qubit/matrix_ops.py:
```diff
@@ -211,25 +226,84 @@ def __init__(
+        **Example**
+
+        >>> U = np.array([[ 0.94877869,  0.31594146], [-0.31594146,  0.94877869]])
+        >>> qml.ControlledQubitUnitary.compute_matrix(U, control_wires=[1], wires=[0], control_values="1")
```
I find it funny that this op has separate `control_wires`, whereas other ops don't; the control wire is inferred from the order.

mariaschuld replied:
Yes that's true! But I guess that's how the CNOT representation is canonically done (control on first), while this gate is something we invent and where the control position is variable...

josh146 commented on pennylane/ops/qubit/matrix_ops.py:
```diff
@@ -211,25 +226,84 @@ def __init__(
+
+        return qml.operation.expand_matrix(
+            canonical_matrix, wires=self.wires, wire_order=wire_order
+        )
```
do you also need to invert the matrix here if the op is inverted?

mariaschuld replied:
I didn't do it because the original method didn't...

josh146 replied:
oh lol!

josh146 commented on pennylane/ops/qubit/matrix_ops.py:
```diff
@@ -211,25 +226,84 @@ def __init__(
+         [ 0.        +0.j -0.31594146+0.j  0.        +0.j  0.94877869+0.j]]
+        """
+        if self._cu is None:
             interface = qml.math.get_interface(self.U)
             left_pad = qml.math.cast_like(qml.math.eye(self._padding_left, like=interface), 1j)
             right_pad = qml.math.cast_like(qml.math.eye(self._padding_right, like=interface), 1j)
-            self._CU = qml.math.block_diag([left_pad, self.U, right_pad])
+            self._cu = qml.math.block_diag([left_pad, self.U, right_pad])
```
we can remove this by simply adding a cache decorator to the `compute_matrix` method no? (albeit, we would need a custom key to deal with the NumPy array, but doable: https://cachetools.readthedocs.io/en/stable/)

josh146 replied:
Actually, there is a potential issue here --- what happens if `U` is mutated during the operations lifetime? the wrong matrix will be returned!

I'm not sure if this would ever happen (and tbh it shouldn't be a thing that's allowed), _but_ it could happen to operations with trainable parameters

mariaschuld replied:
Hmm...that's true. I am happy to just compute this from scratch each time and solve speed by caching some time!

I wasn't sure if the original developer had some ideas here.

What do you think of this option: init computes CU, which is then fed as a param to the super method? We may not even have to overwrite the matrix method in that case because `compute_matrix` just gets CU through `params` and passes it again. Downside is that we carry around a larger object.

josh146 replied:
Sounds good to me!

josh146 commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -705,23 +835,35 @@ class SISWAP(Operation):
+
+        **Example**
+
+        >>> qml.SISWAP.compute_matrix()
```
so many ops I didn't realize we had!

josh146 commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -989,26 +1163,62 @@ def __init__(
+    def compute_matrix(control_int, n_wires):
+        """Canonical matrix representation of the MultiControlledX operator.
+
+        **Examples:**
```
```suggestion
        **Examples**
```
(haven't been noticing, so not sure if the other examples above also have the colon!)

mariaschuld replied:
By now I should have learnt it, this was an outlier I hope

josh146 commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -989,26 +1163,62 @@ def __init__(
+         [0. 0. 1. 0.]
+         [0. 0. 0. 1.]]
+
+        Args:
```
Don't forget to move the example below `Returns`!

josh146 commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -989,26 +1163,62 @@ def __init__(
+
+    def matrix(self, wire_order=None):
+
+        if self._cx is None:
+            # store matrix, so we do not have to do this each time
+            self._cx = self.compute_matrix(self.control_int, len(self.wires))
+
+        canonical_matrix = self._cx
```
I think this might suffer from the same problem as above; the stored canonical matrix won't be returned if `control_int` or the wires change.

Which I think is solved using the cache decorator approach instead

josh146 commented on pennylane/ops/qubit/parametric_ops.py:
```diff
@@ -590,8 +684,8 @@ class PauliRot(Operation):
     _PAULI_CONJUGATION_MATRICES = {
-        "X": Hadamard._matrix(),
-        "Y": RX._matrix(np.pi / 2),
+        "X": Hadamard.compute_matrix(),
+        "Y": RX.compute_matrix(np.pi / 2),
```
this looks so much nicer

josh146 commented on pennylane/transforms/metric_tensor.py:
```diff
@@ -1,641 +1,641 @@
-            return _aux
-
-    return tape.num_wires
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
```
For some reason this entire file looks different? Has anything changed here?

mariaschuld replied:
Oh no, this must be a rendering issue, I should only have updated names (if at all)

josh146 commented on tests/devices/test_default_mixed.py:
```diff
@@ -255,9 +255,9 @@ class TestKrausOps:
-        (PauliX, np.array([[0, 1], [1, 0]])),
-        (Hadamard, np.array([[INV_SQRT2, INV_SQRT2], [INV_SQRT2, -INV_SQRT2]])),
-        (CNOT, np.array([[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 0, 1], [0, 0, 1, 0]])),
+        (PauliX(wires=0), np.array([[0, 1], [1, 0]])),
```
wait, how come this has changed?

mariaschuld replied:
Because then we can just do `matrix()` below, which is much more convenient than `compute_matrix`

josh146 commented on tests/test_operation.py:
```diff
@@ -1506,3 +1505,97 @@ def test_composed(self):
+
+        # the entry should propagate from position (0, 1) in the original tensor
+        expected = tf.constant([[0.0, 1.0], [0.0, 0.0]])
+        assert np.allclose(gradient, expected, atol=tol)
```
Tests look really thorough! A very minor suggestion, I wonder if you copy and paste the `utils.expand` tests here, if they will pass + maybe cover additional edge cases?

josh146 reviewed (commented):
Very thorough PR, and huge effort to add all the docstrings! Happy to approve once the comments/questions are resolved

mariaschuld commented on pennylane/operation.py:
```diff
@@ -129,6 +129,100 @@
+
+    .. note::
+
+        This function has essentially the same behaviour as ``pennylane.utils.expand`` but is fully
+        differentiable.
```
Uh, that is me not having tested this statement to be 100% true :)

mariaschuld commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -1018,46 +1192,46 @@ def __init__(
-        return self._CX
+        Args:
+           control_wires (Iterable): wires to place controls on
+           control_values (str): string of bits determining the controls
```
I changed it to fit the usual pattern. These are part of the hyperparameters of `MultiControlledX`. It has no trainable params.

mariaschuld commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -1070,7 +1244,32 @@ def _parse_control_values(control_wires, control_values):
-        return control_int
+        padding_left = control_int * 2
+        padding_right = 2 ** (len(control_wires)+1) - 2 - padding_left
+        cx = block_diag(np.eye(padding_left), PauliX.compute_matrix(), np.eye(padding_right))
```
The final matrix gets computed here from scratch in every call.

josh146 replied:
:+1:

mariaschuld commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -1070,7 +1244,32 @@ def _parse_control_values(control_wires, control_values):
+    def matrix(self, wire_order=None):
+
+        canonical_matrix = self.compute_matrix(self.hyperparameters["control_wires"],
+                                               self.hyperparameters["control_values"])
```
Much cleaner: the main computation is done in `compute_matrix`

mariaschuld commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -1070,7 +1244,32 @@ def _parse_control_values(control_wires, control_values):
+
+    @property
+    def control_wires(self):
+        return self.hyperparameters["control_wires"]
```
I'm not sure we actually need this, I think it's for the circuit drawer. But in principle, the hyperparams now contain this information. Maybe delete in params/hyperparams refactor.

josh146 replied:
I _think_ it makes sense to keep this if it is an important API for other parts of the codebase

mariaschuld commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -1083,35 +1282,40 @@ def adjoint(self):
+        control_wires = self.hyperparameters["control_wires"]
+        control_values = self.hyperparameters["control_values"]
+        work_wires = self.hyperparameters["work_wires"]
+        target_wire = self.hyperparameters["target_wire"]
```
I'm now using the hyperparams in the decomposition as well...

mariaschuld commented on pennylane/drawer/representation_resolver.py:
```diff
@@ -157,7 +157,7 @@ def _format_controlled_qubit_unitary(operation, symbol, cache, offset):
             str: The formatted operation
         """
-        mat = operation.U
+        mat = operation.parameters[0]
```
avoids saving the unitary specifically for the drawer

mariaschuld replied:
( Updated the drawer accordingly)

mariaschuld commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -1018,46 +1192,48 @@ def __init__(
+            "work_wires": work_wires,
+            "control_wires": control_wires,
+            "control_values": control_values,
+        }
```
Same changes as in `ControlledQubitUnitary`: We save the hyperparameters and then compute matrices in the matrix methods...

mariaschuld commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -1018,46 +1192,48 @@ def __init__(
+        total_wires = control_wires + wires
 
-        super().__init__(*params, wires=wires, do_queue=do_queue)
+        super().__init__(wires=total_wires, do_queue=do_queue)
```
Actually this op does not have trainable params, good to make this explicit!

josh146 replied:
nice!

mariaschuld commented:
@josh146 @albi3ro this is ready and polished for final review.

@josh146 I implemented your suggestions. Major changes since your last review are:

* I reverted back to the `NotImplementedError` system as per our discussion. I will do this also for all other PRs and the dev branch. Since it is the current solution and maybe the best future one, it's best to keep for now.
* I removed developer notes from the `matrix` docstrings and will mention stuff in the developer guide.
* There were two ops where I overwrote `matrix` to do some logic that was partially stored in `self`, `ControlledQubitUnitary` and `MultiControlledX`. Following your suggestion, I moved the entire matrix computation to `compute_matrix`, so I do not have to overwrite `matrix` any more. To do this, I implemented "hyperparameters" for these gates already. 

I found that overall the system of parameters and hyperparameters is quite powerful. If the matrix does not depend on both, they should simply be falsely defined (since we want to make it a convention that parameters + hyperparameters fully define the transformation, while wires defines the subspace it acts on). Can't wait to have it implemented!

**Last remaining open question:**

The above change uncovered something that puzzles me: In `MultiControlledX` one hyperparameter, the wire determining where the X is, was not ever used to construct the matrix. Is the existing logic (which I merely reproduce) wrong? Or is it not a true hyperparameter somehow?

mariaschuld commented on pennylane/operation.py:
```diff
@@ -300,48 +394,92 @@ def hash(self):
+        if wire_order is None or self.wires == Wires(wire_order):
+            return canonical_matrix
+
+        return expand_matrix(canonical_matrix, wires=self.wires, wire_order=wire_order)
```
I thought more about `qml.matrix(op)`, and I am not 100% sure we should do it. Users should be able to get info from an operator without knowing some other methods defined in other modules! I think this would be quite important...

josh146 replied:
I think my motivation here is 99% for the documentation. If we have `qml.matrix(op)` there is _one_ main docstring with canonical examples, description, gotchas, etc.

In this approach however, there are many different `SomeOperator.matrix` that show up in the docs and google results that need to be maintained, found by the user, etc:

![image](https://user-images.githubusercontent.com/2959003/146742959-16bf9a27-c12a-49df-890e-fbe253663da2.png)

josh146 replied:
I often think of this as well:
https://github.com/numpy/numpy/issues/7452

> If we could go back in time then I'd probably argue for not having these methods on ndarray at all, and using functions exclusively. In python, methods are a closed set of operations that have to be specified at class creation time, can't be extended, and can be overridden by subclasses; functions are an open set of operations that can be added to at any time, namespaced, and can't be overridden by subclasses (without some special work anyway). In the ndarray case, it's turned out that method subclassing is ineffective, and that there are many many operations with new ones added all the time, that these can't be written in terms of a small finite set of core operations, and thus that we need to solve the problem of overriding functions for new types anyway. So this all argues for using functions exclusively (sorry if you find it less readable). Too bad we don't have a time machine :-)

josh146 replied:
this is not a comment on this PR, just a general thought!

mariaschuld commented on pennylane/ops/qubit/matrix_ops.py:
```diff
@@ -183,24 +198,45 @@ def __init__(
+         [ 0.        +0.j  1.        +0.j  0.        +0.j  0.        +0.j]
+         [ 0.        +0.j  0.        +0.j  0.94877869+0.j  0.31594146+0.j]
+         [ 0.        +0.j  0.        +0.j -0.31594146+0.j  0.94877869+0.j]]
+        """
```
This looks horrendous, but I basically just moved the entire logic from ``init``

josh146 replied:
:+1:

mariaschuld commented on pennylane/ops/qubit/matrix_ops.py:
```diff
@@ -250,11 +264,23 @@ def _parse_control_values(control_wires, control_values):
+
+    @property
+    def control_wires(self):
+        return self.hyperparameters["control_wires"]
```
We can probably remove this going forward, I think it is used in the drawer.

albi3ro replied:
and circuit compilation.

josh146 replied:
Wait, why should we remove this?

> and circuit compilation.

Yep, in fact @rmoyard has added `target_wires` as well in #1712

mariaschuld commented on tests/ops/qubit/test_parametric_ops.py:
```diff
@@ -1283,9 +1329,26 @@ class TestMultiRZ:
+                [0.0 + 0.0j, 0.0 + 0.0j, 0.0 + 0.0j, 0.99875026 - 0.04997917j],
+            ]
+        )
+
```
Codecov complained, because we overwrite `matrix` and don't test the non-trivial case.

mariaschuld commented on tests/templates/test_subroutines/test_qft.py:
```diff
@@ -104,3 +104,30 @@ def test_QFT_adjoint_decomposition(self, n_qubits):  # tol
+        assert np.allclose(res_static, expected, atol=tol, rtol=0)
+        assert np.allclose(res_dynamic, expected, atol=tol, rtol=0)
+
+        expected_permuted = [
```
Need to test this because we overwrite `matrix`, otherwise codecov complains

mariaschuld commented on tests/test_operation.py:
```diff
@@ -1506,3 +1511,242 @@ def test_composed(self):
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_expand_matrix_usage_in_operator_class(self, tol):
```
Added this test for codecov

mariaschuld commented:
the failing test came about when I merged master and op-refactor...we should fix it there, so for now ignore on this branch please

josh146 commented on pennylane/operation.py:
```diff
@@ -129,6 +129,100 @@
+    """Re-express a base matrix acting on a subspace defined by a set of wire labels
+    according to a global wire order.
+
+    .. note::
+
+        This function has essentially the same behaviour as :func:`.utils.expand` but is fully
+        differentiable.
+
```
Minor: before merging, could you also add a similar `.. note::` to the `utils.expand` docstring?

josh146 commented on pennylane/operation.py:
```diff
@@ -300,48 +394,92 @@ def hash(self):
-    def _matrix(cls, *params):
-        """Matrix representation of the operator
-        in the computational basis.
+    # pylint:disable=unused-argument
+    @staticmethod
+    def compute_matrix(*params, **hyperparams):
```
I mentioned this in #2045 just now, but I have a feeling that putting this here simply applies the pylint comment to the entire class? Would be safer to do

```python
def compute_matrix(*params, **hyperparams):
    # pylint:disable=unused-argument
```

mariaschuld replied:
I tried to find all other instances too!

josh146 commented on pennylane/ops/qubit/matrix_ops.py:
```diff
@@ -183,24 +198,45 @@ def __init__(
-        self._target_wires = wires
-        self._control_wires = control_wires
-        self.U = U
+        self._hyperparameters = {
+            "u_wires": wires,
+            "control_wires": control_wires,
+            "control_values": control_values,
+        }
```
I really like this

josh146 replied:
I like how it provides a place for developers to add hyperparameters, but I also like how there is no restriction on the key names. I think this is important - saying 'you have to name the keys this' and then relying on these names for built-in logic is too much overhead for developers and contributors

josh146 replied:
Everything that _is_ needed for internal logic should be a well-documented property/method instead

josh146 commented on pennylane/ops/qubit/non_parametric_ops.py:
```diff
@@ -1018,46 +1192,54 @@ def __init__(
 
-        control_int = self._parse_control_values(control_wires, control_values)
-        self.control_values = control_values
+        # we do not store this in the hyperparameters,
+        # because it influences the implementation of this op,
+        # not the transformation itself
+        self.work_wires = work_wires
```
not 100% sure I follow this comment?

mariaschuld replied:
I think Christina adds this to the hyperparams. But it is not needed to represent the matrix, so in my books it is not a hyperparameter :) But maybe I have to widen my definition a bit, since it steers the decomposition...

josh146 commented on pennylane/templates/subroutines/grover.py:
```diff
@@ -135,15 +135,20 @@ def expand(self):
-    @classmethod
-    def _matrix(cls, *params):
-        num_wires = params[0]
+    def matrix(self, wire_order=None):
+
+        # note: compute_matrix requires the number of wires, which is why we need to overwrite this method
+        canonical_matrix = self.compute_matrix(len(self.wires))
+
+        if wire_order is None or self.wires == qml.wires.Wires(wire_order):
+            return canonical_matrix
+
+        return qml.operation.expand_matrix(
+            canonical_matrix, wires=self.wires, wire_order=wire_order
+        )
+
+    @staticmethod
+    def compute_matrix(num_wires):
```
can this one also take the `hyperparams` approach?

josh146 commented on pennylane/templates/subroutines/qft.py:
```diff
@@ -70,17 +70,22 @@ def circuit_qft(basis_state):
-    def matrix(self):
-        # Redefine the property here to allow for a custom _matrix signature
-        mat = self._matrix(len(self.wires))
+    def matrix(self, wire_order=None):
```
same question here :)

josh146 reviewed (approved):
Looks good, happy to approve! A huge amount of work 💯 

Only things I think to do before merging is just fixing some of the pylint comments, and also I think some `n_wires` ops were missed?

albi3ro commented on pennylane/operation.py:
```diff
@@ -300,48 +394,91 @@ def hash(self):
-        new operations and observables, that returns the matrix representing
-        the operator in the computational basis.
+        The canonical matrix is the textbook matrix representation that does not consider wires.
+        Implicitly, this assumes that the wires of the operator correspond to the glabal wire order.
```
```suggestion
        Implicitly, this assumes that the wires of the operator correspond to the global wire order.
```

albi3ro commented on pennylane/ops/qubit/matrix_ops.py:
```diff
@@ -183,24 +198,47 @@ def __init__(
-        self._target_wires = wires
-        self._control_wires = control_wires
-        self.U = U
+        self._hyperparameters = {
+            "u_wires": wires,
+            "control_wires": control_wires,
+            "control_values": control_values,
+        }
```
```suggestion
        self.hyperparameters["u_wires"] = wires
        self.hyperparameters["control_wires"] = control_wires
        self.hyperparameters['control_values'] = control_values
```
Another way to potentially do this.

albi3ro reviewed (approved):
🚀
